File: frontend/src/api/ranking.ts
import { queryOptions } from "@tanstack/react-query";

import { api, unwrap } from "@/api/client";
import type { components } from "@/api/schema";
import type { Stage } from "@/api/screening";

export type RankingStatus = components["schemas"]["RankingStatus"];
export type RecallCurve = components["schemas"]["RecallCurve"];
export type Curve = components["schemas"]["Curve"];
export type StoppingAdvice = components["schemas"]["StoppingAdvice"];

export const rankingKeys = {
  all: (pid: string) => ["projects", pid, "ranking"] as const,
  status: (pid: string, stage: Stage) => ["projects", pid, "ranking", "status", stage] as const,
  curve: (pid: string, stage: Stage) => ["projects", pid, "ranking", "curve", stage] as const,
  // Under "screening", so that it is refreshed with the progress after decisions.
  stopping: (pid: string, stage: Stage) =>
    ["projects", pid, "screening", "stopping", stage] as const,
};

export const rankingStatusQuery = (pid: string, stage: Stage) =>
  queryOptions({
    queryKey: rankingKeys.status(pid, stage),
    queryFn: async ({ signal }) =>
      unwrap(
        await api.GET("/api/v1/projects/{pid}/ranking/status", {
          signal,
          params: { path: { pid }, query: { stage } },
        }),
      ),
  });

export const recallCurveQuery = (pid: string, stage: Stage) =>
  queryOptions({
    queryKey: rankingKeys.curve(pid, stage),
    queryFn: async ({ signal }) =>
      unwrap(
        await api.GET("/api/v1/projects/{pid}/ranking/curve", {
          signal,
          params: { path: { pid }, query: { stage } },
        }),
      ),
  });

export const stoppingQuery = (pid: string, stage: Stage) =>
  queryOptions({
    queryKey: rankingKeys.stopping(pid, stage),
    queryFn: async ({ signal }) =>
      unwrap(
        await api.GET("/api/v1/projects/{pid}/screening/stopping", {
          signal,
          params: { path: { pid }, query: { stage } },
        }),
      ),
  });

export async function trainRanking(pid: string, stage: Stage) {
  return unwrap(
    await api.POST("/api/v1/projects/{pid}/ranking/train", {
      params: { path: { pid }, query: { stage } },
    }),
  );
}

File: backend/app/api/v1/ranking.py
"""Relevance ranking: the model's state, retraining on demand, the recall curve and the
stopping-rule helper (guide 8.5, 8.10, 10).

Team-wide numbers are shown only to someone allowed to see other people's decisions
(guide 8.6); the services decide that.
"""

from typing import Annotated

from fastapi import APIRouter, Query

from app.api.deps import RankingDep, ReviewerAccess, ViewerAccess
from app.api.responses import PROJECT
from app.models import ScreeningStage
from app.schemas.ranking import RankingStatus, RecallCurve, StoppingAdvice, TrainStarted

router = APIRouter(prefix="/projects/{pid}", tags=["ranking"])

Stage = Annotated[ScreeningStage, Query()]


@router.get("/ranking/status", responses=PROJECT)
async def ranking_status(
    access: ViewerAccess, ranking: RankingDep, stage: Stage = ScreeningStage.TITLE_ABSTRACT
) -> RankingStatus:
    """Whether a model exists for the stage, and what it has learnt from."""
    return await ranking.status(access, stage)


@router.post("/ranking/train", status_code=202, responses=PROJECT)
async def ranking_train(
    access: ReviewerAccess, ranking: RankingDep, stage: Stage = ScreeningStage.TITLE_ABSTRACT
) -> TrainStarted:
    """Retrain now rather than after the next 25 decisions (guide 8.10)."""
    return await ranking.train(access, stage)


@router.get("/ranking/curve", responses=PROJECT)
async def recall_curve(
    access: ViewerAccess, ranking: RankingDep, stage: Stage = ScreeningStage.TITLE_ABSTRACT
) -> RecallCurve:
    """Includes found against records screened: mine, and the team's if I may see it."""
    return await ranking.curve(access, stage)


@router.get("/screening/stopping", responses=PROJECT)
async def stopping_advice(
    access: ReviewerAccess, ranking: RankingDep, stage: Stage = ScreeningStage.TITLE_ABSTRACT
) -> StoppingAdvice:
    """Guide 8.5: how many records in a row I have excluded, and, once that reaches the
    review's rule, an estimate of the relevant records still unscreened. Advice only."""
    return await ranking.stopping(access, stage)
